calico/felix/plugins/fiptgenerator.py

```diff
@@ -783,70 +784,96 @@ def _rule_to_iptables_fragments_inner(self, chain_name, rule, ip_version,
-                assert isinstance(icmp_code, int), "ICMP code should be an int"
-                icmp_filter = "%s/%s" % (icmp_type, icmp_code)
-            else:
-                icmp_filter = icmp_type
-            if proto == "icmp" and ip_version == 4:
-                append("--match icmp", "--icmp-type", icmp_filter)
-            elif ip_version == 6:
-                assert proto == "icmpv6"
-                # Note variant spelling of icmp[v]6
-                append("--match icmp6", "--icmpv6-type", icmp_filter)
+
+        for neg_pfx in ["", "!"]:
+            maybe_neg_proto = rule.get(neg_pfx + "protocol")
+            if maybe_neg_proto:
+                append(neg_pfx, "--protocol", str(maybe_neg_proto))
+
+            for dirn in ["src", "dst"]:
+                # Some params use the long-form of the name.
+                direction = "source" if dirn == "src" else "destination"
+
+                # Network (CIDR).
+                net_key = neg_pfx + dirn + "_net"
+                if net_key in rule and rule[net_key] is not None:
+                    ip_or_cidr = rule[net_key]
+                    if (":" in ip_or_cidr) == (ip_version == 6):
+                        append(neg_pfx, "--%s" % direction, ip_or_cidr)
+
+                # Tag, which maps to an ipset.
+                tag_key = neg_pfx + dirn + "_tag"
+                if tag_key in rule and rule[tag_key] is not None:
+                    ipset_name = tag_to_ipset[rule[tag_key]]
+                    append("--match set",
+                           neg_pfx, "--match-set", ipset_name, dirn)
+
+                # Selector, likewise.
+                sel_key = neg_pfx + dirn + "_selector"
+                if sel_key in rule and rule[sel_key] is not None:
+                    ipset_name = selector_to_ipset[rule[sel_key]]
+                    append("--match set",
+                           neg_pfx, "--match-set", ipset_name, dirn)
+
+                # Port lists/ranges, which we map to multiport.
+                ports_key = neg_pfx + dirn + "_ports"
+                ports = rule.get(ports_key)
+                if ports:  # Ignore empty list.
+                    # Can only match if the (non-negated) is set to a supported
+                    # value.
+                    assert proto in ["tcp", "udp"], \
+                        "Protocol %s not supported with %s (%s)" % \
+                        (proto, ports_key, rule)
+                    if neg_pfx == '':
+                        # Positive match; caller has already chunked the
+                        # ports list up into blocks of suitable length for
+                        # _ports_to_multiport.  We only see one chunk.
+                        ports_str = self._ports_to_multiport(ports)
+                        append("--match multiport", "--%s-ports" % direction,
+                               ports_str)
+                    else:
+                        # This is a negative match.  While an individual
```

Comment:
We don't really need to support this.  We could also support positive >15 matches if we wanted, by adding more rules in.  It probably won't scale that far anyway - I imagine there'll be a limit to how many different matches we can use anyway.